### ISSUE TITLE
custom usercode with lpoptions isn't shown in Libreoffice

### ISSUE THREAD

**Issue by TxOfficial** [OpenPrinting/cups#405]
Today i check the behaviour on the new Ubuntu 22.04 with CUPS 2.4.1 and the following PPD:

https://www.openprinting.org/download/PPD/Ricoh/PS/Ricoh-IM_C3000_PS.ppd

When I use the print dialog from libre office and setup the usercode manually within the appplication printdialog,  I got the attached PS-file "01_Print_via_libre_print_dialog_usercode_0850.ps.txt" within the UserCode 0850. Then I closed libre office, used lpoptions on the command line (refer screenshot 05_lpoptions_command.png), started libre again and print without any changes in the printdialog of libre I got "02_Print_via_libre_lpoption.ps.txt" with an empty UserCode. At this time lpoption -l and cat ~/.cups/lpoption has the values from the attached Screenshot "03_Values of lpoption.png". See the asterisk (*) in the usercode line at "Custom.STRING". When I open the printdialog in libre and check the usercode option the option was also set to custom but with an empty value (not "0850") and the printresult is wrong as "02_...ps.txt". I've checked the cups error_log, but can't find any entry for UserCode "0850" at this time and also the processing of the lpoptions command after I setup  "cupsctl --debug-logging" takes a really long time, approx. 20sec after the cursor/promt comes back. Interesting is, when i use the lp-command "lp -d print2welp -o UserCode=Custom.0850 test.pdf" it works and I get a correct file similar to "01_....ps.txt". So i think there is a bug within the lpoptions command

Hope you can find anything that can help us in the attached cups error_log. Until approx. 1,1MB you find the first bad printout (lpoption) without success and after that the second one with the successful lp command.

[01_Print_via_libre_print_dialog_usercode_0850.ps.txt](https://github.com/OpenPrinting/cups/files/8799172/01_Print_via_libre_print_dialog_usercode_0850.ps.txt)
[02_Print_via_libre_lpoption.ps.txt](https://github.com/OpenPrinting/cups/files/8799173/02_Print_via_libre_lpoption.ps.txt)
![03_Values of lpoption](https://user-images.githubusercontent.com/106541893/171014834-6e8b5536-3054-475e-abd6-8a769378ddcd.png)
![04_libre print dialog after set of lpoption - empty value but switched to custom](https://user-images.githubusercontent.com/106541893/171014839-35da6f1b-56e8-4217-a8b7-00164c4c393f.png)
[error_log.txt](https://github.com/OpenPrinting/cups/files/8799174/error_log.txt)
![05_lpoptions_command](https://user-images.githubusercontent.com/106541893/171017589-a4c4c67b-a197-4635-822a-439d5756ca2a.png)

**Comment by zdohnal** [comment]
Hi @TxOfficial ,

thank you for reporting the issue!

Unfortunately it looks you've forgot to attach the .txt file with debug logs - I see only the links to PPD, postscript outputs and screenshots - it would be great if you attached the debug log file as well.

Ad the issue - since you change a default value via lpoptions, AFAIU it should be used during normal ```lp``` invocation (without -o). Do you get the correct output that way?

**Comment by TxOfficial** [comment]
Hi @zdohnal,

the "error_log.txt" is still there, between the last two screenshots. I think that the lp commands works but we want to use lpoptions to change the default when we print with the application and not via lp. If we choose one of the predefind values 1001,1002, or 1003 it works also with the application but not with UserCode=Custom.xxxx. Hope you can find and fix the problem?

Many thanks and best regards..
Tx

**Comment by zdohnal** [comment]
> the "error_log.txt" is still there, between the last two screenshots.

Aha, completely missed it - I was looking among the other .txt files, but didn't look between the screenshots...

> I think that the lp commands works but we want to use lpoptions to change the default when we print with the application and not via lp. 

I've asked about ```lp``` command to narrow the issue - whether there is something completely wrong with lpoptions, or is there a problem with how LibreOffice uses CUPS API to do this.

So it would be great if you tried to print via lp without -o option - if lp correctly loads ```Custom.0850``` then there will a difference how LibreOffice loads those characters, thus it will need a change in their application/or in our API which it uses.

**Comment by TxOfficial** [comment]
Hi @zdohnal,

yes it works when i use lp without the -o option. 

Any idea how to fix this with the CUPS API that libre use?

**Comment by zdohnal** [comment]
Ok, I've checked the code which LibreOffice uses and compared with ```lp``` command and CUPS API documentation. LibreOffice uses ```cupsGetDests()``` function, which is useful for listing printers, but once user selects a specific printer, ```cupsGetNamedDest()``` function has to be used for getting the full option set, altogether with local option overrides such as your UserCode defined by lpoptions.

From cups.dest.c:

```
1727  * 'cupsGetNamedDest()' - Get options for the named destination.
1728  *
1729  * This function is optimized for retrieving a single destination and should
1730  * be used instead of @link cupsGetDests2@ and @link cupsGetDest@ when you
1731  * either know the name of the destination or want to print to the default
1732  * destination.  If @code NULL@ is returned, the destination does not exist or
1733  * there is no default destination.
```

So I cannot fix this in CUPS - the proper fix has to happen in LibreOffice, so the app can use the correct function and get users the correct options.

It would be great if you reported this issue to Libreoffice project.

Thank you in advance!


Zdenek

**Comment by zdohnal** [comment]
@michaelrsweet ok, then I will reinvestigate - thanks for checking!

**Comment by zdohnal** [comment]
Ok, I was able to reproduce this with CUPS 2.4.2 and Libreoffice 7.3.3.2. I set Custom.0850 via lpoptions and lp command is able to apply the option into job, but Libreoffice doesn't.

Then I wrote a simple program with ```cupsGetDests2()```, looked for ricoh printer and went through its options in gdb. Among the options was:

```
(gdb) p dests[6].options[22]
$31 = {name = 0x41b144 "UserCode", value = 0x405fc4 "Custom.0850"}
```

CUPS correctly passes the user defined value for UserCode via ```cupsGetDests2()```, so the issue is in Libreoffice about not applying/filtering out this option.
